internal/feature_test.go

```diff
@@ -132,106 +142,134 @@ func TestScriptBasedCore(t *testing.T) {
 	}
 	defer feature.Disconnect(true)
 
-	testDownloadInstall(feature, mc, w.GenerateSoftwareArtifacts(false, "install"), t)
-
-	feature.serverCert = testCert
-	testDownloadInstall(feature, mc, wSecure.GenerateSoftwareArtifacts(true, "install"), t)
-}
-
-func testDownloadInstall(feature *ScriptBasedSoftwareUpdatable, mc *mockedClient, artifacts []*hawkbit.SoftwareArtifactAction, t *testing.T) {
-	// Preapare simple software update action.
-	sua := &hawkbit.SoftwareUpdateAction{
-		CorrelationID: "test-correlation-id",
-		SoftwareModules: []*hawkbit.SoftwareModuleAction{{
-			SoftwareModule: &hawkbit.SoftwareModuleID{Name: "test", Version: "1.0.0"},
-			Artifacts:      artifacts,
-			Metadata:       map[string]string{"artifact-type": "plane"},
-		}},
+	if basic {
+		testDownloadInstall(feature, mc, w.GenerateSoftwareArtifacts(false, "install"), t)
+		feature.serverCert = testCert
+		testDownloadInstall(feature, mc, wSecure.GenerateSoftwareArtifacts(true, "install"), t)
+	} else {
+		testResume(feature, mc, w.GenerateSoftwareArtifacts(false, "install"), t)
 	}
+}
 
-	// 2. Try to execute simple an download operation.
-	feature.downloadHandler(sua, feature.su)
+func testResume(feature *ScriptBasedSoftwareUpdatable, mc *mockedClient, artifacts []*hawkbit.SoftwareArtifactAction, t *testing.T) {
+	sua := prepareSoftwareUpdateAction(artifacts)
+	testReconnectWhileRunningOperation(feature, mc, sua, false, t) // disconnect while downloading
+	testReconnectWhileRunningOperation(feature, mc, sua, true, t)  // disconnect while downloading or installing
+}
 
-	// 2.1. Check for STARTED status.
-	if lo := mc.lastOperation(t); lo["status"] != "STARTED" {
-		t.Fatalf("received unexpected lastOperation status: %s != STARTED", lo["status"])
+func testReconnectWhileRunningOperation(feature *ScriptBasedSoftwareUpdatable, mc *mockedClient, sua *hawkbit.SoftwareUpdateAction, install bool, t *testing.T) {
+	var waitDisconnect sync.WaitGroup
+	waitDisconnect.Add(1)
+	go func() {
+		if install {
+			feature.installHandler(sua, feature.su)
+		} else {
+			feature.downloadHandler(sua, feature.su)
+		}
+		feature.Disconnect(false)
+		waitDisconnect.Done()
+	}()
+
+	statuses := pullStatusChanges(mc)
+	waitDisconnect.Wait()
+
+	// reconnecting
+	feature.Connect(mc, supConfig, edgeCfg)
+	statuses = append(statuses, pullStatusChanges(mc)...)
+	if install {
+		checkInstallStatusEvents(statuses, t)
+	} else {
+		checkDownloadStatusEvents(statuses, t)
 	}
+}
 
-	// 2.2. Check for DOWNLOADING status.
-	if lo := mc.lastOperation(t); lo["status"] != "DOWNLOADING" {
-		t.Fatalf("received unexpected lastOperation status: %s != DOWNLOADING", lo["status"])
+func pullStatusChanges(mc *mockedClient) []interface{} {
+	var statuses []interface{}
+	for i := 0; i < 50; i++ {
```

Comment:
this hardcoded magic number is not very strict, do we know the number? is it fixed or variable?

Reply:
It is used as a some kind of limitation/timeout